lib/puppet/transaction/report.rb

```diff
@@ -308,6 +308,16 @@ def initialize_from_hash(data)
     end
   end
 
+  def calculate_resource_status_tuple(key, rs)
+    if rs.nil?
+      [key, nil]
+    elsif !rs.out_of_sync? && Puppet[:exclude_unchanged_resources]
```

Comment:
I'm having second thoughts about calling `out_of_sync?` because it defaults to `false`

https://github.com/puppetlabs/puppet/blob/e95386bbcb2b4eefb898fd83ac47eaec3bbeb021/lib/puppet/resource/status.rb#L162

And only changes to `true` if at least one event has been added:

https://github.com/puppetlabs/puppet/blob/e95386bbcb2b4eefb898fd83ac47eaec3bbeb021/lib/puppet/resource/status.rb#L131

But if a child resource (file `/a/b`) depends on a parent resource (dir `/a`) and the parent fails, then the child will be `skipped` without ever being evaluated:

https://github.com/puppetlabs/puppet/blob/e95386bbcb2b4eefb898fd83ac47eaec3bbeb021/lib/puppet/transaction.rb#L282

Similarly a resource may be marked as `failed` without ever being evaluated if its provider never becomes suitable during the run:

https://github.com/puppetlabs/puppet/blob/e95386bbcb2b4eefb898fd83ac47eaec3bbeb021/lib/puppet/transaction.rb#L134

So we may want to only exclude resources that were scheduled (meaning we tried to evaluate them) and their events array is empty:

https://github.com/puppetlabs/puppet/blob/e95386bbcb2b4eefb898fd83ac47eaec3bbeb021/lib/puppet/transaction.rb#L287

Reply:
Hmm, I'm not seeing `scheduled` as a consistent resource attribute in my debugging, so that may not work; I need to redo my test vms because I have too much debugging strewn about, but I will link some more docs about what I am seeing in the `last_run_report.yaml`.

Reply:
Sounds good. I'd expect there to be an accessor for each STATE:

https://github.com/puppetlabs/puppet/blob/e95386bbcb2b4eefb898fd83ac47eaec3bbeb021/lib/puppet/resource/status.rb#L33-L34

And then there's some metaprogramming to create a predicate method for each of those, e.g. `scheduled?`

https://github.com/puppetlabs/puppet/blob/e95386bbcb2b4eefb898fd83ac47eaec3bbeb021/lib/puppet/resource/status.rb#L110-L114

Reply:
ohhh, `scheduled` is never serialized in `Status#to_data_hash`